File: esw-ocs/esw-ocs-dsl-kt/src/main/kotlin/esw/ocs/dsl/highlevel/AlarmServiceDsl.kt
package esw.ocs.dsl.highlevel

import csw.alarm.api.javadsl.IAlarmService
import csw.alarm.models.AlarmSeverity
import csw.alarm.models.Key.AlarmKey
import kotlinx.coroutines.delay
import kotlinx.coroutines.future.asDeferred
import kotlinx.coroutines.future.await
import kotlinx.coroutines.launch
import kotlin.time.Duration

interface AlarmServiceDsl : LoopDsl {
    val alarmService: IAlarmService
    val _alarmRefreshDuration: Duration

    companion object {
        private val keyToSeverityMap: HashMap<AlarmKey, AlarmSeverity> = HashMap()
    }

    /**
     * Sets alarm severity against provided alarm key and keeps refreshing it after every `csw-alarm.refresh-interval` which by default is 3 seconds
     *
     * @param alarmKey unique alarm in alarm store e.g nfiraos.trombone.tromboneaxislowlimitalarm
     * @param severity severity to be set for the alarm e.g. Okay, Warning, Major, Critical, etc
     */
    suspend fun setSeverity(alarmKey: AlarmKey, severity: AlarmSeverity) {
        keyToSeverityMap[alarmKey] = severity
        alarmService.setSeverity(alarmKey, severity).await()
        if (keyToSeverityMap.size == 1)
            delayTaskExecution(_alarmRefreshDuration) { startSetSeverity() } // alarm is already set at this stage, hence wait for refresh interval and then start loop
    }

    private fun delayTaskExecution(delayDuration: Duration, task: () -> Unit) = coroutineScope.launch {
        delay(delayDuration.toLongMilliseconds())
        task()
    }

    private fun startSetSeverity() = loopAsync(_alarmRefreshDuration) {
        keyToSeverityMap.map { (key, severity) -> alarmService.setSeverity(key, severity).asDeferred() }
    }
}

File: esw-ocs/esw-ocs-dsl-scala/src/main/scala/esw/ocs/dsl2/highlevel/AlarmServiceDsl.scala
package esw.ocs.dsl2.highlevel

import csw.alarm.api.scaladsl.AlarmService
import csw.alarm.models.AlarmSeverity
import csw.alarm.models.Key.AlarmKey
import esw.ocs.dsl.script.StrandEc

import scala.concurrent.duration.Duration
import async.Async.*
import scala.concurrent.{ExecutionContext, Future}

class AlarmServiceDsl(alarmService: AlarmService, _alarmRefreshDuration: Duration, loopDsl: LoopDsl)(using ExecutionContext):
  private var keyToSeverityMap: Map[AlarmKey, AlarmSeverity] = Map()

  inline def setSeverity(alarmKey: AlarmKey, severity: AlarmSeverity): Unit = await(setSeverityAsync(alarmKey, severity))

  private def setSeverityAsync(alarmKey: AlarmKey, severity: AlarmSeverity): Future[Unit] = async {
    keyToSeverityMap += (alarmKey -> severity)
    await(alarmService.setSeverity(alarmKey, severity))
    if (keyToSeverityMap.size == 1)
      // alarm is already set at this stage, hence wait for refresh interval and then start loop
      await {
        delayTaskExecution(_alarmRefreshDuration) {
          startSetSeverity()
        }
      }
  }

  private def delayTaskExecution(delayDuration: Duration)(task: => Unit): Future[Unit] = async {
    loopDsl.delay(delayDuration)
    task
  }

  private def startSetSeverity() = loopDsl.loopAsync(_alarmRefreshDuration) {
    val futures = keyToSeverityMap.map { (key, severity) =>
      alarmService.setSeverity(key, severity)
    }
    await(Future.sequence(futures))
  }

end AlarmServiceDsl
